RFS/src/main/java/org/opensearch/migrations/bulkload/workcoordination/OpenSearchWorkCoordinator.java

```diff
@@ -645,6 +665,145 @@ private WorkItemAndDuration getAssignedWorkItem(LeaseChecker leaseChecker,
+
+        var createResults = new ArrayList<Boolean>();
+        resultTree.path("items").elements().forEachRemaining(
+                item -> createResults.add(acceptableStatusCodes.contains(item.path("create").path("status").asInt()))
+        );
+        if (createResults.contains(false)) {
+            throw new IllegalStateException(
+                    "One or more of the successor work item(s) could not be created: "
+                            + String.join(", ", workItemIds)
+                            + ".  Response: "
+                            + response.toDiagnosticString()
+            );
+        }
+
+    }
+
+    @Override
+    public void createSuccessorWorkItemsAndMarkComplete(
+            String workItemId,
+            ArrayList<String> successorWorkItemIds,
+            Supplier<IWorkCoordinationContexts.ICreateSuccessorWorkItemsContext> contextSupplier
+    ) throws IOException, InterruptedException, IllegalStateException {
+        if (successorWorkItemIds.contains(workItemId)) {
+            log.atError().setMessage("successorWorkItemIds {} can not not contain the parent workItemId: {}").addArgument(successorWorkItemIds).addArgument(workItemId).log();
+            throw new IllegalArgumentException("successorWorkItemIds can not not contain the parent workItemId");
+        }
+        try (var ctx = contextSupplier.get()) {
+            int updateWorkItemRetries = 0;
+            while (true) {
+                try {
+                    updateWorkItemWithSuccessors(workItemId, successorWorkItemIds);
+                    break;
+                } catch (IllegalStateException e) {
+                    updateWorkItemRetries++;
+                    if (updateWorkItemRetries > MAX_CREATE_SUCCESSOR_WORK_ITEMS_RETRIES) {
+                        ctx.addTraceException(e, true);
+                        log.atError().setCause(e).setMessage(
+                                "Throwing exception because max tries (" + MAX_CREATE_SUCCESSOR_WORK_ITEMS_RETRIES + ")" +
+                                        " have been exhausted").log();
+                        throw new RetriesExceededException(e, updateWorkItemRetries);
+                    }
+                    ctx.addTraceException(e, true);
+                    var sleepBeforeNextRetryDuration = Duration.ofMillis(
+                                    (long) (Math.pow(2.0, (updateWorkItemRetries-1)) * CREATE_SUCCESSOR_WORK_ITEMS_RETRY_BASE_MS));
+                    ctx.addTraceException(e, false);
+                    log.atWarn().setCause(e)
+                            .setMessage("Couldn't complete work assignment due to exception. Backing off {} and trying again.")
+                            .addArgument(sleepBeforeNextRetryDuration).log();
+                    Thread.sleep(sleepBeforeNextRetryDuration.toMillis());
+                }
+            }
+
+            createUnassignedWorkItemsIfNonexistent(successorWorkItemIds);
```

Comment:
I don't think it should continue indefinitely here. The first step is the critical one. As long as that first step has finished, the next worker to pick this up will just pick up where it left off in trying to create the successorWorkItems. This process is expected to be run from the `exitOnLeaseTermination` function, so it has a limited amount of time to work with before another worker will pick up the lease. If the cluster is running very slowly or with a high error rate, it's possible/likely that another worker will pick up this lease during the retries and attempt to create these successor work items at the same time. Due to the structure of the bulk request, that won't actually cause any problems, but it's extra work for both the workers and the cluster.

I think it's safer to give this a small number of attempts (I'm leaning towards one, but open to maybe 3?) and then move on so the task gets picked up by a different worker later.

Reply:
Take a look at the counts for the top, I agree that it doesn't need to be indefinite since we'll die and another process can eventually drive. this to completion, but there IS a minor cost to not doing the splits ASAP (since workers can't grab the new work items, they might be spinning down & take longer to come back up, etc).